### DISCUSSION TITLE
Add global context

### DISCUSSION THREAD

**Discussion by notramo** [martenframework/marten#227] [Ideas]
With `isolated_inclusions` enabled, it's tedious to pass some variables around, e.g. CSRF token. I recommend adding a `global_context` which is similar to `context`, but it is available everywhere, regardless of whether isolated inclusions are enabled or disabled. Context producers should put the produced values there, and it should be available to handlers too, in a similar way as `context`.

**Comment by ellmetha** [comment]
I am not sure to be convinced. If a particular inclusion requires the presence of "global context" variables to function properly, perhaps it's an indication that this included template shouldn't be isolated in the first place. 🤷‍♂️

In my view, complicating the process of setting context variables only to accommodate isolated inclusions doesn't seem justified. Especially considering that the defaults of the Marten framework are (i) no strict variables and (ii) no isolated inclusions. I don't think it's warranted to introduce this level of complexity to all the users of the framework given that most of those will simply use the defaults, which are simple enough and work for most cases.

To me, the solution to what you are describing would be to use contextual inclusions.

**Reply by notramo** [reply]
Contextual inclusions can be messy. In most cases, isolated inclusions result in cleaner code, however, variables added by context producers mostly need to be accessed from arbitrary locations, which don't really correlate with the context. Most frequent example is the CSRF token if it's added to the context via `before_render` in the base class.

This can be compared to SvelteKit where components mostly receive data via attributes, however, the `$page` (and other stores) can be imported into any component without having to pass it explicitly as attribute. Also, the `setContext` mechanism is another example of not having to pass data explicitly to every child from every parent.

What about making context producer variables unaffected by context isolation?

**Reply by ellmetha** [reply]
Variables provided by context producers are added to all rendered templates, except included templates (unless they are contextual of course). Let's say I have a template in which I include another template that way:

```jinja
{% include "partials/my_snippet.html" with foo="bar" isolated %}
```

The context available to the `partials/my_snippet.html` should comply with what was defined in the including template and only contain a `foo` variable. If we let the included template have access to context producer variables, we are not respecting what's defined in the template and the included template is not really isolated.

To me, and given this example, it does not appear consistent to automatically include context producer variables in included templates. As a user, I would be surprised that the framework does not respect what I am defining and that my included templates are not isolated completely. The fact that context producers define variables that are added to all templates does not mean that those should be added to included templates if they are supposed to be isolated.

Maybe having isolated templates that have access to context producer variables would work in your specific project case, but I am not sure this would apply to all projects out there, to be honest. As for limitations concerning the utilization of CSRF tokens in isolated templates, alternative solutions could likely be explored to mitigate such issues.

**Reply by ellmetha** [reply]
FYI regarding the use of the `csrf_token` tag in isolated templates, this is now possible (this was fixed in d33487b).
